document a proposed code review process (#247)
mattermost/mattermost-developer-documentation: site/content/contribute/getting-started/code-review.md

---
title: "Code Review"
date: 2018-03-06T00:00:00-04:00
weight: 5
subsection: Getting Started
---

All changes to the product must be reviewed.

* User experience changes must be reviewed by a [product manager](/contribute/getting-started/core-committers/#product-managers).
* Code changes must be reviewed by at least two [core committers](/contribute/getting-started/core-committers/#core-committers).

> * Documentation changes must be reviewed by a [product manager](/contribute/getting-started/core-committers/#product-managers).

Author: Hmm, I'm actually unsure what "editor review" even means for this repo. I'd always assumed that was @jasonblais :D. @jasonblais, can you clarify what `Editor Review` means for this repo and possibly others?

  * Product managers may ask for reviews from [core committers](/contribute/getting-started/core-committers/#core-committers) and [QA testers](/contribute/getting-started/core-committers/#qa-testers) as required.
* Any change may be verified by a [QA tester](/contribute/getting-started/core-committers/#qa-testers) as required.

If you are a community member seeking a review
----------------------------------------------

1. Submit your pull request.
    * Follow the [contribution checklist](https://developers.mattermost.com/contribute/getting-started/contribution-checklist/).

> 2. Reach out to the author of the ticket your PR resolves.

Reviewer: Typically Eric's GitHub account is the author of all the help wanted issues so I'm not sure if this works unless they look at the JIRA ticket as well. Not really sure of a foolproof way for them to know who to mention, unless they've been interacting with a core committer somewhere about the ticket already

Author: Agreed. I've removed the mention of the author, and updated the language here.

    * If you are Mattermost staff, try to take into account the timeoff calendar.
    * Try to avoid assigning the same person to all of your reviews unless they are related.
    * When in doubt, ask for recommendations on our community server.
4. Optionally assign a [QA tester](/contribute/getting-started/core-committers/#qa-testers) and label your pull request with `QA Review`.
    * Not every pull request requires QA Review. Prioritize large features or material changes to the product for QA review.
    * Pull requests labelled with `QA Review` automatically notify the QA team at large.
    * Reviews by QA may occur at the same time as review by core committers.
5. Apply additional labels as necessary:
    * `CherryPick/Approved`: Apply this if the pull request is meant for a quality or patch release.
    * `Do Not Merge/Awaiting PR`: Apply this if the pull request depends on another (e.g. server changes)
    * `Setup Test Server`: Apply this to create a test server with your changes for review.
    * See [labels](/contribute/getting-started/labels) for more details.

> 6. Assign a milestone as necessary.

Reviewer 2: Why would you not assign a milestone? At least for reviewer it's more clear what the expectations are and what the "deadline" is.

Author: My impression is that not everything is "scheduled" for a given release -- sometimes that decision is deferred while a review is ongoing. Omitting a milestone essentially amounts to a signal indicating a somewhat lower priority, though I don't think that changes any of the SLAs around required feedback.

Reviewer 2: Agree. It's better to not create additional pressure for milestone and keep things flexible.

    * Most issues are targeted for an upcoming release, and should be assigned the corresponding milestone.
    * The milestone is mandatory for bug fixes that must be cherry-picked.
7. Wait for a review.
    * Expect some interaction with each of your reviewers within 2 business days.
    * There is no need to explicitly mention them on the pull request or to explicitly reach out on our community server.
    * Core committers and QA testers are expected to have the GitHub plugin installed to automate notifications and to trigger a daily review of their outstanding requested reviews.
8. Make any necessary changes.
    * If a reviewer requests changes, your pull request will disappear from their queue of reviews.

>     * Once you've addressed the concerns, assign them as a reviewer again to put your pull request back on their plate.

Reviewer: ```suggestion
    * Once you've addressed the concerns, assign them as a reviewer again to put your pull request back in their queue.
```

Not sure if non-native English speakers will understand "back on their plate"

9. Merge the pull request.
    * Only merge once all concerns have been addressed, any reviewing product manager has approved the changes, any reviewing QA tester has approved the changes, and at least two core committers have approved.
    * Remove the `1: PM Review`, `2: Dev Revew` and `QA Review` labels and assign the `3: Ready to Merge` label.
    * Merge your pull request and delete the branch if not from a fork.
    * Note that the last core committer to approve your changes may do this on your behalf.
    * If your pull request depends on other pull requests, consider assigning the `Do Not Merge/Awaiting PR` label.
10. Handle any cherry-picks.
    * Cherry-pick the commit from `master` to the affected releases.
    * Push your changes directly to the remote branch.
    * No new pull request is required unless there are substantial merge conflicts.

>     * Remove the `CherryPickApproved` label and apply the `CherryPickDone` label.

Reviewer 2: ```suggestion
    * Remove the `CherryPick/Approved` label and apply the `CherryPick/Done` label.
```

If you are a core committer asked to give a review
--------------------------------------------------

1. Respond promptly to requested reviews.
    * Assume the requested review is urgent and blocking unless explicitly stated otherwise.
    * Try to interact with the author within 2 business days.
    * Configure the GitHub plugin to automate notifications.

>     * Review your outstanding requested reviews daily to avoid blocking authors.

Reviewer 2: :100:
